### PR TITLE
Dev feature roadmap

### PR DESCRIPTION
**Pull request by F-said**
Updated UML to include "write_eeg_data" function. Modified returned and passed objects to be explicit. 

Omitted "return raw" following:
> To keep the diagram explicit, I think we could omit "raw" from the return values of the function calls since the functions are modifying the raw data in place and not returning. Let me know if I'm misunderstanding the pipeline, however.

Will modify UML if the above comment is inaccurate.

UML

### CODE REVIEW

**Comment by F-said** [comment]
Once this is finalized, I think we can begin the `feature-main` branch that will have the skeleton main pipeline code.

This will involve creating:
- the prepared directory structure
Ex: 

```yml
baseEEG
├── scripts
|    ├── run_pipeline.py
|    ├── filter.py
|    ├── bad-channels.py
|    ├── ica.py
|    ├── segment.py
|    ├── final-reject.py
├── tests
├── docker
|    ...
```
- prepared import statements
```python
import user_params_load

import filter
import bad-channels
import ica
import segment
# etc 
```
- pipeline code, which will be non-functioning until feature branches are merged
```python
def main():
    bids_root = pathlib.Path(user_params["root"])
    # get subjects
    for subject in subject_collect:
        # get subject data
        # run pipeline on subject
```

**Comment by georgebuzzell** [comment]
@F-said this is really fantastic! Thank you!

Minor question (from ignorance, not from disagreeing): Can you please ELI5 "raw is modified in place"? And relatedly, why does the UML say that we "pass raw" if raw is "modified in place"? I think I am not clear on a CS use of the terms here and standards of a UML diagram (new to this). Assuming that the way you are describing/depicting things is in line with best practices, then I would not want to change things just so it is more intuitive to my CS-naive mind. But, we will want to create a second diagram in that case that is more intuitive for novices so that they can "see" that raw keeps getting updated as main progresses; does that make sense?

Another minor comment: please also update to account for epochs, per a comment you made elsewhere. 

Yes, agree that once this PR is merged, it would be great to set up main as you suggest. 

Above all else, this is really, really fantastic work @F-said thank you!

**Review by georgebuzzell — COMMENTED** [review]
@F-said this is really fantastic! Thank you!

Minor question (from ignorance, not from disagreeing): Can you please ELI5 "raw is modified in place"? And relatedly, why does the UML say that we "pass raw" if raw is "modified in place"? I think I am not clear on a CS use of the terms here and standards of a UML diagram (new to this). Assuming that the way you are describing/depicting things is in line with best practices, then I would not want to change things just so it is more intuitive to my CS-naive mind. But, we will want to create a second diagram in that case that is more intuitive for novices so that they can "see" that raw keeps getting updated as main progresses; does that make sense?

Another minor comment: please also update to account for epochs, per a comment you made elsewhere.

Yes, agree that once this PR is merged, it would be great to set up main as you suggest.

Above all else, this is really, really fantastic work @F-said thank you!

**Comment by F-said** [comment]
@georgebuzzell 

I think I actually may be misunderstanding the pipeline actually, should pipeline functions be returning copies of the raw mutated data?

For example, in preprocessing_filter we do:
```python
return raw.copy().filter(l_freq=l_freq, h_freq=h_freq)
```

Will this be the standard for all functions?

**Comment by georgebuzzell** [comment]
@F-said The pipeline is a series of steps, where changes are made to the original data (i.e. filtering) and the output (i.e. filtered) data from one step is passed into the next step (i.e. bad chan removal), and the output of that step is in turn passed to the next step. The order of the steps matter, and they "build upon each other" by making further changes to the data. Does that make sense?

I think part of the confusion is that we just keep referring to the data as "raw" which may not be the best name as it is passed along through the pipeline and processed. However, for whatever reason, that seems to be standard naming in men, so I guess we should keep that. 

So, to be clear, the pipeline should load in raw, and load in user_params. Then, raw and user params are passed to the first step (function) that does something to raw, and also makes a copy of raw that is renamed and saved to disk, and also creates an output_dict. Then, the next step in main (function) takes in raw (which was modified by the first step/function) and also takes in user_params, and again modifies raw, and makes another copy of raw that is renamed and written to disk, as well as creates an output dict. This continues for all steps. The copies of raw, which are named something else and written to disk, ARE NOT returned to main, instead, only the raw that was modified by that step, but not renamed is returned.

What is unclear to me, is if we are not renaming raw, but still modifying it, is it not proper to say that raw is "returned" to main from the function/step that modified it?

**Comment by F-said** [comment]
closing PR until overview of pipeline is figured out per #106
